File: 0009. Flood Fill.cpp
// time = O(n)
// space = O(1)

class Solution {
public:
    void fill (vector<vector<int>>& image, int r, int c, int startingColor, int newColor) {
        if (r < 0 || c < 0 || r >= image.size() || c >= image[0].size() || image[r][c] != startingColor) return;
        image[r][c] = newColor;
        for (auto &[x, y]: vector<pair<int, int>>{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}) {
            fill (image, r + x, c + y, startingColor, newColor);
        }
    }
    
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int newColor) {
        if (image[sr][sc] == newColor) return image;
        fill(image, sr, sc, image[sr][sc], newColor);
        return image;
    }
};

File: py3/0009_flood_fill.py
# time: O(n)
# space: O(n)
from typing import List
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        initial_color, frontier = image[sr][sc], []
        if initial_color != color:
            frontier.append((sr, sc))
        while frontier:
            cr, cc = frontier.pop()
            image[cr][cc] = color
            for x, y in ((-1, 0), (1, 0), (0, 1), (0, -1)):
                nr, nc = cr + x, cc + y
                if 0 <= nr < len(image) and 0 <= nc < len(image[0]) and image[nr][nc] == initial_color:
                    frontier.append((nr, nc))
        return image        
